### PR TITLE
Fix compress bound off by one

### PR DESCRIPTION
**Pull request by jlee303**
Fix bug from #1674, thanks @PQCraft for finding bug!

LZ4F_writeOpen() allocates its output buffer using LZ4F_compressBound(0, prefs). However, LZ4F_compressBound(0) returns the bound for LZ4F_flush() and LZ4F_compressEnd() operations. Fix by using LZ4F_compressBound(blockSize, prefs)

Added test case to detect issue:
Here is the error message when using LZ4F_compressBound(0, prefs)
```
LZ4F_compressBound(0) =  65551 
LZ4F_compressBound(1 KB, autoFlush=1) =  1032 
LZ4F_compressBound(1 KB, autoFlush=0) =  65544 
LZ4F_write checksums disabled edge case : Error detected ! 
```

I did not see a test file for lz4file so I added the test case to frametest.c, let me know if I should move the test to a better location!

### CODE REVIEW

File: tests/frametest.c [hunk 1]
```diff
@@ -442,25 +442,28 @@ static int unitTests(U32 seed, double compressibility)
         LZ4F_preferences_t writePrefs;
         size_t             blockSize;
         char*              data;
+        int                failed;
```
Comment (Cyan4973) [inline tests/frametest.c]: minor: 
I would prefer initialization to happen at the same place as declaration.
This would give : `int failed = 0;` rather than 2 separated lines with an uninitialized state in between.

And yes, this comment is valid for all other variables declared in this block.

File: tests/frametest.c [hunk 2]
```diff
@@ -442,25 +442,28 @@ static int unitTests(U32 seed, double compressibility)
...
 
         tmpFile   = tmpfile();
         writeCtx  = NULL;
         blockSize = LZ4F_getBlockSize(LZ4F_default);
         data      = (char*)malloc(blockSize);
+        failed    = 0;
 
         memset(data, 0, blockSize);
```
Comment (Cyan4973) [inline tests/frametest.c]: The main issue, on reaching this code, is that `data` is not tested. So allocation could have failed. In which case, this `memset()` is writing into `NULL`.

2 solutions:
- allocate with `calloc()`, which automatically fills with zero on success
- test and branch out on allocation failure before `memset()`.

Note that `calloc()` spares you the `memset()`, but you'll still need to test and branch out on allocation failure.

File: tests/frametest.c [hunk 3]
```diff
@@ -442,25 +442,28 @@ static int unitTests(U32 seed, double compressibility)
...
         memset(&writePrefs, 0, sizeof(writePrefs));
         writePrefs.frameInfo.contentChecksumFlag = LZ4F_noContentChecksum;
         writePrefs.frameInfo.blockChecksumFlag   = LZ4F_noBlockChecksum;
 
-        if (!tmpFile || !data) goto _output_error;
-        if (LZ4F_isError(LZ4F_writeOpen(&writeCtx, tmpFile, &writePrefs))) goto _output_error;
-        if (LZ4F_isError(LZ4F_write(writeCtx, data, blockSize - 1))) goto _output_error;
-        if (LZ4F_isError(LZ4F_write(writeCtx, data, 1))) goto _output_error;
+        if (!tmpFile || !data) failed = 1;
```
Comment (Cyan4973) [inline tests/frametest.c]: checking `!data` here is already too late.

Also you may not need an additional variable is all you need is to scope out the next lines into a `{ ... }` block.

File: tests/frametest.c [hunk 4]
```diff
@@ -442,25 +442,28 @@ static int unitTests(U32 seed, double compressibility)
...
+        if (!failed && LZ4F_isError(LZ4F_writeOpen(&writeCtx, tmpFile, &writePrefs))) failed = 1;
```
Comment (Cyan4973) [inline tests/frametest.c]: I guess it works, but it feels unusual.

I would prefer a pattern like : 

```
do {
    if (status(operation1(arg1)) == fail) break; /* exit the while loop */

    if (status(operation2(arg2)) == fail) break; 

    // etc...
} while(0);
```

File: tests/frametest.c [hunk 5]
```diff
@@ -51,24 +51,22 @@
...
 /* unoptimized version; solves endianness & alignment issues */
-static void FUZ_writeLE32 (void* dstVoidPtr, U32 value32)
+static void FUZ_writeLE32(void* dstVoidPtr, U32 value32)
 {
     BYTE* const dstPtr = (BYTE*)dstVoidPtr;
-    dstPtr[0] = (BYTE) value32;
-    dstPtr[1] = (BYTE)(value32 >> 8);
-    dstPtr[2] = (BYTE)(value32 >> 16);
-    dstPtr[3] = (BYTE)(value32 >> 24);
+    dstPtr[0]          = (BYTE)value32;
```
Comment (Cyan4973) [inline tests/frametest.c]: minor: it seems this PR has been mixed with a formatting exercise.
It's not in itself bad, but it makes it harder to find the changes that matter, hence more difficult to review.
That's why formatting is typically left as a separate exercise.

File: lib/lz4file.c [hunk 6]
```diff
@@ -275,7 +275,7 @@ LZ4F_errorCode_t LZ4F_writeOpen(LZ4_writeFile_t** lz4fWrite, FILE* fp, const LZ4
   writeFile->maxWriteSize = blockSize;
 
   /* Calculate and allocate destination buffer */
-  writeFile->dstBufMaxSize = LZ4F_compressBound(0, prefsPtr);
+  writeFile->dstBufMaxSize = LZ4F_compressBound(blockSize, prefsPtr);
```
Comment (Cyan4973) [inline lib/lz4file.c:278]: 👍 the change that matters

**Review by Cyan4973 — CHANGES_REQUESTED** [review]
The library change itself is correct.

The focus in more on the test side.
There are a few patterns that deserve some attention.

File: tests/frametest.c (hunk 5 above)
Comment (jlee303) [reply tests/frametest.c]: Reverted formatting changes! Sorry auto format was on 😸

File: tests/frametest.c (hunk 4 above)
Comment (jlee303) [reply tests/frametest.c]: Updated! Lemme know if there's anything else :)

**Review by Cyan4973 — APPROVED** [review]
Great work @jlee303 !
